lib/indexer/indexer.js

```diff
@@ -0,0 +1,690 @@
+        // Make sure that this is the current block.
+        const current = await this.getBlockMeta(this.height);
+        if (current.hash.compare(block.hash()) !== 0)
+          return false;
+
+        await this._removeBlock(meta, block, view);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Sync with the chain.
+   * @private
+   * @returns {Promise}
+   */
+
+  async _syncChain() {
+    let height = this.height;
+
+    // In the case that the indexer has never
+    // started, sync to the best height.
+    if (!height) {
+      await this._rollforward();
+      return;
+    }
+
+    // Check for a re-org that might
+    // leave chain in a different state.
+    // Scan chain backwards until we
+    // find a common height.
+    while (height > 0) {
+      const meta = await this.getBlockMeta(height);
+      assert(meta);
+
+      if (await this.getEntry(meta.hash))
+        break;
+
+      height -= 1;
+    }
+
+    if (height < this.height) {
+      await this._rollback(height);
+      await this._rollforward();
+    } else {
+      await this._rollforward();
+    }
+  }
+
+  /**
+   * Scan blockchain to the best chain height.
+   * @private
+   * @returns {Promise}
+   */
+
+  async _rollforward() {
+    this.logger.info('Indexing to best height from height (%d).', this.height);
+
+    for (let height = this.height + 1; ; height++) {
+      const entry = await this.getEntry(height);
+      if (!entry)
+        break;
+
+      const meta = new BlockMeta(entry.hash, height);
+
+      const block = await this.chain.getBlock(entry.hash);
+      assert(block);
+
+      const view = await this.chain.getBlockView(block);
+      assert(view);
+
+      await this._addBlock(meta, block, view);
+    }
+  }
+
+  /**
+   * Rollback to a given chain height.
+   * @param {Number} height
+   * @returns {Promise}
+   */
+
+  async _rollback(height) {
+    if (height > this.height) {
+      this.logger.warning(
+        'Ignoring rollback to future height (%d).',
+        height);
+      return;
+    }
+
+    this.logger.info('Rolling back to height %d.', height);
+
+    while (this.height > height && this.height > 1) {
+      const meta = await this.getBlockMeta(this.height);
+      assert(meta);
+
+      const block = await this.chain.getBlock(meta.hash);
+      assert(block);
+
+      const view = await this.chain.getBlockView(block);
+      assert(view);
+
+      await this._removeBlock(meta, block, view);
+    }
+  }
+
+  /**
+   * Add a block's transactions without a lock.
+   * @private
+   * @param {BlockMeta} meta
+   * @param {Block} block
+   * @param {CoinView} view
+   * @returns {Promise}
+   */
+
+  async _addBlock(meta, block, view) {
+    assert(block.hasRaw(), 'Expected raw data for block.');
+
+    const start = util.bench();
+
+    if (meta.height !== this.height + 1)
+      throw new Error('Indexer: Can not add block.');
+
+    // Start the batch write.
+    this.start();
```

Comment:
Even though this is fine, passing batch objects explicitly makes review of those functions much easier. Also it makes those functions harder to misuse. E.g. if somehow indexBlock was used incorrectly (before start()) it will throw an error. (`this.put` will fail because `this.batch` will be undefined).

Also makes batching harder, you can't easily compose batches.

Reply:
The method `indexBlock()` is called within `_addBlock()`, which is called from syncing related functions, and `indexBlock()` shouldn't be used otherwise (aside from testing). It's adding to an atomic commit that includes `_setTip()`. This is very similar to the atomic updates for `lib/blockchain/chaindb.js`, and shares a similar interface.

Reply:
yeah, I saw them. 
I just think batching state hides important details,
I just think the future changes or others trying to implement on top could get in some trouble with this pattern.
Testing can catch most of the bugs that could occur.